### PR TITLE
Updated ApplicationDeployer to conditionally delete published applica…

### PR DESCRIPTION
**Pull request by kichalla**
…tion folder

@JunTaoLuo

### CODE REVIEW

**Comment by Tratcher** [comment]
Why?
Also, why make it an environment variable rather than a DeploymentParameter?

**Comment by kichalla** [comment]
> Why?
Also, why make it an environment variable rather than a DeploymentParameter?

Basically when a test fails/succeeds we currently delete the published folder which prevents from diagnosing the issue if I want to look at the published output and then run the app. I chose environment variable as other approaches (ex: deployment parameter) would require code to be changed to start debugging.

Yeah, I know that having an environment variable reduces discoverability but this would be the same with other approaches too?

**Comment by Tratcher** [comment]
Are there any other places in the deployers that directly read environment variables? I'd still favor code-first.

**Comment by kichalla** [comment]
No, there are no other places reading environment variables.

**Comment by JunTaoLuo** [comment]
Instead of reading the environment variable in the deployers, maybe have the test projects read/parse it and then pass it along as a deployment parameter?

**Comment by kichalla** [comment]
Yeah, I considered that but since all test projects need to do this wanted to avoid it. Anyways, I guess I will just use deployment parameters and each test project can read the environment variable and set it as needed.

File: src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs [hunk 1]
```diff
@@ -82,14 +82,31 @@ protected void DotnetPublish(string publishRoot = null)
 
         protected void CleanPublishedOutput()
         {
-            try
+            var deletePublishedFolderEnvironmentVariableName = "DELETE_PUBLISHED_FOLDER";
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs]: Would also recommend to add a prefix `ASPNETCORE_` just to differentiate.

File: src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs [hunk 2]
```diff
@@ -82,14 +82,31 @@ protected void DotnetPublish(string publishRoot = null)
...
+            var deleteLocallyPublishedFolder = true;
+            var setting = Environment.GetEnvironmentVariable(deletePublishedFolderEnvironmentVariableName);
+            if (string.Equals(setting, "false", StringComparison.OrdinalIgnoreCase))
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs]: I would take either `false` or `0` to disable the setting. See https://github.com/aspnet/Hosting/blob/41b4e6017e5d2c67e3a3f031558ded23abd28d58/src/Microsoft.AspNetCore.Hosting/Internal/WebHostOptions.cs#L45-L49.

**Comment by JunTaoLuo** [comment]
The rest of the change looks good, some minor comments on the parsing of environment variables that will be moved to the tests.

**Comment by kichalla** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs [hunk 3]
```diff
@@ -82,14 +82,23 @@ protected void DotnetPublish(string publishRoot = null)
...
+                catch (Exception exception)
+                {
+                    Logger.LogWarning($"Failed to delete directory : {exception.Message}");
+                }
             }
-            catch (Exception exception)
+            else
             {
-                Logger.LogWarning($"Failed to delete directory : {exception.Message}");
+                Logger.LogWarning(
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.Testing/Deployers/ApplicationDeployer.cs]: should be `is set to 'false'` right?

**Comment by JunTaoLuo** [comment]
Changes looks good but the exception message looks wrong. With that fixed: :shipit: .

File: src/Microsoft.AspNetCore.Server.Testing/Common/DeploymentParameters.cs [hunk 4]
```diff
@@ -81,6 +81,8 @@ public class DeploymentParameters
         /// </summary>
         public bool PublishApplicationBeforeDeployment { get; set; }
 
+        public bool DeletePublishedApplicationOnDispose { get; set; } = true;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Testing/Common/DeploymentParameters.cs]: How about PreservePublishedApplicaitonForDebugging = false?

Comment (kichalla) [inline src/Microsoft.AspNetCore.Server.Testing/Common/DeploymentParameters.cs]: Sounds good. I already pushed my changes but will submit one more commit.

**Pull request merged**